src/LanguageServer/Protocol/Handler/InlayHint/InlayHintResolveHandler.cs

```diff
@@ -33,39 +34,58 @@ public TextDocumentIdentifier GetTextDocumentIdentifier(LSP.InlayHint request)
         public Task<LSP.InlayHint> HandleRequestAsync(LSP.InlayHint request, RequestContext context, CancellationToken cancellationToken)
         {
             var document = context.GetRequiredDocument();
-            return ResolveInlayHintAsync(document, request, _inlayHintCache, cancellationToken);
+            var options = globalOptions.GetInlineHintsOptions(document.Project.Language);
+            var inlayHintCache = context.GetRequiredService<InlayHintCache>();
+            var resolveData = GetInlayHintResolveData(request);
+            return ResolveInlayHintAsync(document, request, inlayHintCache, resolveData, options, cancellationToken);
         }
 
-        internal static async Task<LSP.InlayHint> ResolveInlayHintAsync(Document document, LSP.InlayHint request, InlayHintCache inlayHintCache, CancellationToken cancellationToken)
+        internal static async Task<LSP.InlayHint> ResolveInlayHintAsync(
+            Document document,
+            LSP.InlayHint request,
+            InlayHintCache inlayHintCache,
+            InlayHintResolveData resolveData,
+            InlineHintsOptions options,
+            CancellationToken cancellationToken)
         {
-            var resolveData = GetInlayHintResolveData(request);
-            var (cacheEntry, inlineHintToResolve) = GetCacheEntry(resolveData, inlayHintCache);
-
             var currentSyntaxVersion = await document.GetSyntaxVersionAsync(cancellationToken).ConfigureAwait(false);
-            var cachedSyntaxVersion = cacheEntry.SyntaxVersion;
+            var resolveSyntaxVersion = resolveData.SyntaxVersion;
 
-            if (currentSyntaxVersion != cachedSyntaxVersion)
+            if (currentSyntaxVersion.ToString() != resolveSyntaxVersion)
             {
-                throw new LocalRpcException($"Cached resolve version {cachedSyntaxVersion} does not match current version {currentSyntaxVersion}")
+                throw new LocalRpcException($"Request resolve version {resolveSyntaxVersion} does not match current version {currentSyntaxVersion}")
                 {
                     ErrorCode = LspErrorCodes.ContentModified
                 };
             }
 
-            var taggedText = await inlineHintToResolve.GetDescriptionAsync(document, cancellationToken).ConfigureAwait(false);
+            var inlineHintToResolve = GetCacheEntry(resolveData, inlayHintCache);
+            if (inlineHintToResolve is null)
+            {
+                // It is very possible that the cache no longer contains the hint being resolved (for example, multiple documents open side by side).
+                // Instead of throwing, we can recompute the hints since we've already verified above that the version has not changed.
```

Comment:
aside: if we're recomputing, it seems odd to see so much code here.  as opposed to calling into some common helper that we execute for the normal "give me inlay hints" codepath.